psl-core/src/main/java/org/linqs/psl/reasoner/sgd/SGDReasoner.java

```diff
@@ -60,60 +88,143 @@ public double optimize(TermStore baseTermStore) {
-                        iteration, objective, movement, (end - start), totalTime);
+                log.trace("Iteration {} -- Objective: {}, Normalized Objective: {}, Mean Movement: {}, Iteration Time: {}, Total Optimization Time: {}",
+                        iteration, objective, objective / termCount, meanMovement, (end - start), totalTime);
             }
 
             iteration++;
             termStore.iterationComplete();
-
-            if (breakOptimization(iteration, objective, oldObjective, movement)) {
-                break;
-            }
         }
 
-        log.info("Optimization completed in {} iterations. Objective: {}, Total Optimiztion Time: {}",
-                iteration - 1, objective, totalTime);
-        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termStore.size());
+        objective = computeObjective(termStore);
+        log.info("Optimization completed in {} iterations. Objective: {}, Normalized Objective: {}, Total Optimization Time: {}",
+                iteration, objective, objective / termCount, totalTime);
+        log.debug("Optimized with {} variables and {} terms.", termStore.getNumVariables(), termCount);
 
         termStore.syncAtoms();
 
         return objective;
     }
 
-    private boolean breakOptimization(int iteration, double objective, double oldObjective, float movement) {
+    private float minimize(SGDObjectiveTerm term, float[] variableValues, int iteration) {
+        float movement = 0.0f;
+        float variableStep = 0.0f;
+        float newValue = 0.0f;
+        int[] variableIndexes = term.getVariableIndexes();
+
+        term.computeGradient(variableValues, gradient);
+
+        for (int i = 0; i < term.size(); i++) {
+            if (coordinateStep) {
+                term.computeGradient(variableValues, gradient);
+            }
+            variableStep = computeVariableStep(variableIndexes[i], iteration);
+
+            newValue = Math.max(0.0f, Math.min(1.0f, variableValues[variableIndexes[i]] - variableStep));
+            movement += Math.abs(newValue - variableValues[variableIndexes[i]]);
+            variableValues[variableIndexes[i]] = newValue;
+        }
+
+        return movement;
+    }
+
+    private float computeVariableStep(int variableIndex, int iteration) {
+        float beta1 = 0.9f;
+        float beta2 = 0.999f;
+        float mean_hat = 0.0f;
+        float variance_hat = 0.0f;
+        float step = 0.0f;
+        float adaptedLearningRate = 0.0f;
+
+        if (adaGrad) {
+            accumulatedGradientSquares.put(variableIndex, accumulatedGradientSquares.getOrDefault(variableIndex, 0.0f)
+                    + (float)Math.pow(gradient.get(variableIndex), 2.0f));
+            adaptedLearningRate = calculateAnnealedLearningRate(iteration) / (float)Math.sqrt(accumulatedGradientSquares.get(variableIndex) + 1e-8f);
```

Comment:
I am assuming for this and adam you are hoping learningrateSchedule to be set to CONSTANT. I am not able to see it though. Better to explicit.